pkg/apis/lighthouse.submariner.io/v1alpha1/types.go

```diff
@@ -0,0 +1,66 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+Copyright Contributors to the Submariner project.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ServiceImportWeights is the Schema for the serviceimportweights API.
+type ServiceImportWeightMap struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ServiceImportWeightMapSpec `json:"spec,omitempty"`
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// ServiceImportWeightsSpec defines the desired state of ServiceImportWeights.
+type ServiceImportWeightMapSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Weights map of [src_cluster -> [namespace -> [svc_name -> [target_cluster -> weight]]]]
+	SourceClusterWeightMap map[string]*ClusterWeightMap `json:"source_cluster_weight_map,omitempty"`
+}
```

Comment:
The nesting levels seem overly complex. I think we can simplify by defining a single struct:

```
type WeightSpec struct {
    SourceCluster `json:"sourceCluster"`
    Service `json:"service"`
    Namespace `json:"namespace"`
    TargetCluster `json:"targetCluster"`
    Weight int64 `json:"weight"`
}

Weights []WeightSpec `json:"weights,omitempty"`
```

Is this intended to be distributed or defined per cluster? If the latter then we can drop the `SourceCluster`.

Reply:
You are right about the sourceCluster, left overs of my optimization code - where I build the full map - but no need to pass all the information to all the clusters just their local one.
About the new struct - mapping prevents searching - don't you think it worth using a map?
I thought of maybe having the key to be like a path/dns  `cluster(/.)scv(/.).namespace` = weight
Which will make the nesting simpler - wdyt? @tpantelis

Reply:
> About the new struct - mapping prevents searching - don't you think it worth using a map?

"_mapping prevents searching_" - I'm not sure what you mean here. 

> I thought of maybe having the key to be like a path/dns cluster(/.)scv(/.).namespace = weight
> Which will make the nesting simpler - wdyt?

I'm not clear on why it needs to be specifically keyed and what's the use case for searching in that manner. To me, this CRD is just a vehicle for a user to specify weights per service and cluster.

Another approach is to have the CRD define a single weight policy instead of a singleton list, ie

```
type ServiceWeightPolicy struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec ServiceWeightPolicySpec `json:"spec,omitempty"`
}

type ServiceWeightPolicySpec struct {
    Service `json:"service"`
    Namespace `json:"namespace"`
    TargetCluster `json:"targetCluster"`
    Weight int64 `json:"weight"`
}
```

I think this more standard in K8s.

Reply:
OK so you're referring to the controller implementation wrt searching. Remember this is the user-facing API and should be designed as such and not based on the backend controller implementation. You can build your own `map[string]int64` from each `ServiceWeightPolicy` keyed by `Service + "/" + Namespace + "/" + TargetCluster` similarly as is done in _lighthouse/pkg/serviceimport/controller.go_.

Reply:
Yea you are right, the thing is that calculating the weights is done all at once, and I thought of distributing them once per cluster in one structure to avoid sending a lot of smaller pieces of information for each update (in systems with a lot of services it could result in a lot of API calls).
Do you think it is over-optimizing for the moment? And once we face performance issues we aggregate the calls? Maybe there are APIs of Kubernetes I'm not aware of that could solve this foreseen issue?

Reply:
> Yea you are right, the thing is that calculating the weights is done all at once

I'm not sure what you mean by this...

> , and I thought of distributing them

I'm not sure distributing them would make sense...  but, if so, I wouldn't imagine there would be enough of them to cause a performance hit.